build.wake

```diff
@@ -0,0 +1,430 @@
+  )
+  | distinctBy scmp
+
+def copyDir srcDir destDir fileInputs =
+  def dirInputs =
+    getDestDirs srcDir destDir fileInputs
+    | map mkdir
+  def inputs = dirInputs ++ fileInputs
+  def from_ = simplify "{srcDir}/.."
+  def srcs =
+    fileInputs
+    | map (
+      _
+      | getPathName
+      | from_.relative
+      | simplify
+    )
+    | filter (! matches `\.\./.*` _)
+    | sortBy (_ <~ _)
+    | filterDirs
+  def cmdline =
+    "rsync", "--inplace", "--relative",
+    "--ignore-times", srcs ++ (relative from_ destDir, Nil)
+  makePlan cmdline inputs
+  | setPlanDirectory from_
+  | runJob
+  | getJobOutputs
+
+def filterWakeAndDotFiles =
+  _
+  | filter (!matches `(.*/)*\..*` _.getPathName)
+  | filter (!matches `.*\.wake` _.getPathName)
+
+#
+# Plan Resources
+#
+
+global def defaultFreedomESDKResources =
+  "riscv-tools/2019.08.0",
+  "python/python/3.7.1",
+  Nil
+
+#
+# Virtualenv Creation
+#
+# Freedom E SDK depends on a number of Python libraries. Here we enable
+# the installation of a single, workspace-wide virtualenv which can be
+# preinstalled using either the `preinstallFreedomESDKVirtualenv` target
+# or the global `preinstall` topic. If the virtualenv has not been
+# preinstalled, it will be created on-demand when needed during the
+# creation of a `FreedomESDKTarget`.
+#
+
+global def freedomESDKBuildDir = "build/freedom-e-sdk"
+global def freedomESDKVenvDir = "{freedomESDKBuildDir}/venv"
+
+def makeFreedomESDKVirtualenvJob Unit =
+  def runDir = here
+
+  def inputs =
+    source "freedom-e-sdk/scripts/virtualenv.mk",
+    source "freedom-e-sdk/requirements.txt",
+    mkdir freedomESDKVenvDir,
+    Nil
+
+  def cmd =
+    def venvDir = relative runDir freedomESDKVenvDir
+    "bash", "-c", "%
+      export FREEDOM_E_SDK_VENV_PATH=$(pwd)/%{venvDir}
+      make -f scripts/virtualenv.mk virtualenv
+      # Remove e.g. .fuse/31337 from the paths in the virtualenv
+      find ${FREEDOM_E_SDK_VENV_PATH}/bin -type f | xargs sed -i 's%.fuse/[[:digit:]]\+/%%g'
+      # Remove e.g. .build/31337 from the paths in the virtualenv (when Wake doesn't use FUSE)
+      find ${FREEDOM_E_SDK_VENV_PATH}/bin -type f | xargs sed -i 's%.build/[[:digit:]]\+/%%g'
+      %", Nil
+
+  makePlan cmd inputs
+  | setPlanDirectory runDir
+  | setPlanResources defaultFreedomESDKResources
+  | setPlanKeep True
+
+# preinstallFreedomESDKVirtualenv allows Freedom E SDK users to
+# preinstall the virtualenv in their workspace at the time of their
+# choosing with `wake -x preinstallFreedomESDKVirtualenv Unit`
+global target preinstallFreedomESDKVirtualenv Unit =
+  makeFreedomESDKVirtualenvJob Unit
+  | runJob
+  | getJobStderr
+
+publish preinstall = (preinstallFreedomESDKVirtualenv), Nil
+
+# Wake rules which depend upon the Freedom E SDK virtualenv should
+# add `installFreedomESDKVirtualenv` to their Plan inputs.
+global target installFreedomESDKVirtualenv Unit =
+  makeFreedomESDKVirtualenvJob Unit
+  | runJob
+  | getJobOutputs
```

Comment:
I have started developed a personal pattern/habit of not having any important APIs (e.g. `global` functions) directly return the `getJobOutputs`. There's two specific reasons I have for this:

1. I'm not a fan of the current behavior of `getJobOutputs` returning a list of a single BadPath on error, since it's easy to forget to check for that, and I feel that we often run into confusing errors due to propagating BadPaths too far without handling them. I think @terpstra is general in agreement with me about removing BadPath eventually and replacing it with `Result Path` (where Path is only a "good" path) in most places in the standard library.
2. This makes it difficult/impossible to make backwards compatible changes to this API in the future. For example, if we later decided that we wanted to return an additional piece of information, we can't do so without changing the type signature of this function. If, however, we had this return something like a `PythonVirtualEnv` `tuple`, then we can add new fields to that tuple without breaking any interfaces, since the type signature would continue to work.

We don't necessarily have to go to 2) for all interfaces, but I think all global/exported interfaces are ones where we should strongly consider adding tuples since it's more painful to break global/exported interfaces.

Reply:
Added the `FreedomESDKVirtualenv` tuple.